---
title: "Taxonomic assignment of endophytic isolate E14504F"
author: "Dan Spakowicz"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  word_document: default
  pdf_document:
    citation_package: natbib
    fig_caption: yes
    keep_tex: yes
    latex_engine: xelatex
fontsize: 11pt
geometry: margin=1in
fontfamily: mathpazo
spacing: double
bibliography: e14504f.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE)

# Load the required packages
list.of.packages <- c("seqinr", "googlesheets", "tidyverse", "httpuv", "rentrez",
                      "ape", "treebase", "ips", "ggtree", "ggtree", "Biostrings",
                      "msa", "phangorn", "ggrepel", "phytools")
new.packages <- list.of.packages[!(list.of.packages %in% 
                                     installed.packages()[,"Package"])]

# Install missing packages
if (length(new.packages)) install.packages(new.packages, 
                                           repos = "http://cran.rstudio.com/")
# Load required packages
lapply(list.of.packages, require, character.only = TRUE)

# Set seed
set.seed(12345)
```

Herein are the methods for the phylogenetic analysis of isolate E14504F.

# DNA isolation and sequencing

E14504F was grown on potato dextrose broth for 14 days, after which the mycelium was harvested and the genomic DNA purified using a DNeasy Plant Mini Kit (Qiagen; Hilden, Germany) according to the manufacturerâ€™s protocols. The ITS, LSU and RPB2 loci were amplified using GoTaq Flexi (Promega; Madison, WI) and standard primers (Table S1). Amplicons were sequenced at the W.M. Keck Facility at Yale University in forward and reverse directions. Consensus sequences were generated and manually curated for consistency using the Staden package v2.0.0b8 [@bonfield_new_1995] and were deposited in GenBank.

# Database comparisons

DNA sequences were classified by comparison to several curated online databases, recently reviewed by Prakash et al.[-@prakash_online_2017]. Searching the LSU sequence against the SILVA database returned no taxonomic assignments above confidence threshold. Searching the LSU sequence against the Ribosomal Database Project returned 100% confidence in the genus _Thanatephorus_, within the family Ceratobasidiaceae, order Canthrellales. This genus had been recently circumscribed by Gonzalez et al in an effort to clarify the Rhizoctonia, an anamorphic genus found to be polyphyletic by molecular methods. The BOLD Systems database search of the ITS sequence returned 92% similarity to _Sclerotium hydrophilum_, in the order Agaricales. The UNITE database ITS search resulted in equiprobable matches to _Thanetephorus cucumeris_ as well as another genus in the Ceratobasidiaceae, _Ceratorhiza hydrophila_. To clarify the discrepancy between these databases, representatives from each of the clades were retrieved for single multi-locus phylogenetic analysis.


# Phylogenetic analysis

A phylogenetic tree showing the relationship of E14504F to other fungi was constructed using internal transcribed spacer (ITS), large ribosomal subunit (LSU) and RNA polymerase II (RPB2) sequences (Table S2). Comparison sequences were included if voucher specimens of the organism are available in public repositories and the sequence is associated with a publication. Sequences were aligned using muscle [@edgar_muscle:_2004] within the ape package v4.1 [@paradis_ape:_2004]. Sites with gaps in more than 30% of the alignments were removed. Optimal substitution models were identified for each alignment using modelTest [@darriba_jmodeltest_2012] in the phangorn package v2.2.0 [@schliep_phangorn:_2010]. For each locus, a neighbor-joining tree was optimized by maximum likelihood method using a GTR+G model and then 100 bootstrap replicates performed. The loci were then interleaved and a tree contructed using RAxML v8.2.10 [@stamatakis_raxml_2014-1] and visualized using ggtree v1.8.2 [@yu_ggtree:_2017]. All analyses were performed with R v3.4.2 [@r_foundation_for_statistical_computing_vienna_austria._r_2017] in RStudio v1.1.383 [@rstudio_team_rstudio:_2016]. All sequences used in the study, as well as the individual locus and interleaved trees, are available through TreeBase (treebase.org) XXXXXXX. The code used to generate the trees is available at https://github.com/dspak/e14504f. 

# Code

```{r read fastas as a list of strings}
files <- list.files(path = "../E14504F sequences/", 
                    pattern = "*.fasta", full.names = TRUE)
ofastas <- list()
for (f in files) {
  tmp <- readLines(f)
  tmp[2:length(tmp)] <- tolower(tmp[2:length(tmp)])
  ofastas[[f]] <- tmp
}

# Adjust names of list
CleanPaths <- function(names){
  tmp <- gsub(".*E14504F? (.*)\\.fasta", "\\1", names)
  tmp <- gsub(" partial| FULL| ver2", "\\1", tmp)
  return(tmp)
}
names <- CleanPaths(names(ofastas))
names(ofastas) <- names

# Collapse sequences to one line header, one line sequence
CollapseTwoLines <- function(fastaString){
  if (length(fastaString) > 2) {
    x <- fastaString
    y <- x[2:length(x)]
    y <- paste(y, collapse = "")
    z <- c(x[1], y)
    return(z)
  } else {
    return(fastaString)
  }
}
# Iterate function over the list
for (i in names(ofastas)) {
  ofastas[[i]] <- CollapseTwoLines(ofastas[[i]])
}

# Remove any non-ACTGN nucleotides
ofastas <- lapply(ofastas, function(x) gsub("^[^>](.*)[^actgn]", "\\1n", x))
```

```{r pull sequences from google sheets and retrieve from Genbank}
# Load table into dataframe
sheet <- gs_title("E14504F")
x <- gs_read(sheet)

# Convert hyphen-only columns to NA
x <- data.frame(apply(x, 2, function(x) gsub("^-$", NA, x)), as.is = TRUE)

# Take in a character vector of genbank accession numbers and return a fasta 
# file in ape format
RetrieveSequencesAsString <- function(charvec){
  try({
    string <- entrez_fetch(db = "nucleotide", id = charvec, rettype = "fasta")
    string <- unlist(strsplit(string, split = "\n"))
    return(string)
  }, silent = TRUE)
}

# Retrieve all sequence into a list
loci <- colnames(x[,6:12])
gfastas <- list()
for (i in loci) {
  gfastas[[i]] <- RetrieveSequencesAsString(x[,grep(i, colnames(x))])
}
# Remove those without any sequences
gfastas <- gfastas[-(which(sapply(gfastas, class) == "try-error"))]

# Convert all to lowercase (needed for nucleotide substitution matrix)
# and remove any nucleotides other than actg or n
seqs <- list()
ConvertToLowercase_actgn <- function(gfastas) {
  seqs <- lapply(gfastas, function(x) grep("^[^>]", x))
  for (i in 1:length(gfastas)) {
    for (j in 1:length(seqs[[i]])) {
      tmp <- tolower(gfastas[[i]][seqs[[i]][j]])
      tmp <- unlist(strsplit(tmp, split = ""))
      notn <- grep("[^actgn]", tmp)
      tmp[notn] <- "n" 
      tmp <- paste(tmp, collapse = "")
      gfastas[[i]][seqs[[i]]][j] <- tmp
    }
  }
  return(gfastas)
}
gfastas <- ConvertToLowercase_actgn(gfastas)
```

```{r change all names}
# Create tip labels
tips <- paste(x$Name, x$Voucher.ID, sep = " ")

# Get names for each of the sequences as list
hasseq <- apply(x[6:12], 2, function(x) !is.na(x))
newlabels <- apply(hasseq, 2, function(x) tips[x])

# Reduce labels set to those in gfastas
newlabels <- newlabels[names(newlabels) %in% names(gfastas)]

# Changes all names in gfastas
for (i in 1:length(gfastas)) {
  temp <- tempfile()
  write(gfastas[[i]], temp)
  tmp2 <- seqinr::read.fasta(file = temp)
  temp <- tempfile()
  seqinr::write.fasta(sequences = tmp2, names = newlabels[[i]], file.out = temp)
  gfastas[[i]] <- readLines(temp)
}
```

```{r reverse complement all sequences}
# Read in each string to ape as a fasta file
grfastas <- list()
for (i in 1:length(gfastas)) {
  temp <- tempfile()
  write(gfastas[[i]], temp)
  grfastas[[i]] <- ape::read.FASTA(file = temp)
}
names(grfastas) <- names(gfastas)

# Reverse complement all sequences
grfastas <- lapply(grfastas, ape::complement)

# Write each to fasta and read back in as string
for (i in names(grfastas)) {
  ape::write.dna(x = grfastas[[i]], temp, format = "fasta")
  grfastas[[i]] <- readLines(temp)
}
```

```{r merge e14504f fastas with genbank forward and reverse complements}
# Create a list for "both fastas" and append the genbank string onto Nneoma's
bfastas <- list()
for (n in names(gfastas)){
  if (n %in% names(ofastas)) {
    bfastas[[n]] <- c(ofastas[[which(names(ofastas) %in% n)]], gfastas[[n]])
  } else {
    bfastas[[n]] <- gfastas[[n]]
  }
}

brfastas <- list()
for (n in names(grfastas)){
  if (n %in% names(ofastas)) {
    brfastas[[n]] <- c(ofastas[[which(names(ofastas) %in% n)]], grfastas[[n]])
  } else {
    brfastas[[n]] <- grfastas[[n]]
  }
}
```

## Alignments for distance checking

```{r alignment each locus}
#### ALERT !!!
# This section takes ~ 10 minutes to run
# Set to eval=FALSE by default
# 
# Read in each string as a fasta file
fastas <- list()
for (i in 1:length(bfastas)) {
  temp <- tempfile()
  write(bfastas[[i]], temp)
  fastas[[i]] <- ape::read.FASTA(file = temp)
}

# Align sequences
alns <- lapply(fastas, ape::muscle)
names(alns) <- names(bfastas)

# Read in each string as a fasta file
rfastas <- list()
for (i in 1:length(brfastas)) {
  temp <- tempfile()
  write(brfastas[[i]], temp)
  rfastas[[i]] <- ape::read.FASTA(file = temp)
}

# Align sequences
ralns <- lapply(rfastas, ape::muscle)
```

```{r score the alignments}
# Distance calculation for each alignment
fdist <- lapply(alns, function(x) dist.dna(x, as.matrix = TRUE))
rdist <- lapply(ralns, function(x) dist.dna(x, as.matrix = TRUE))
names(fdist) <- names(rdist) <- names(gfastas)

# Retrieve the E14504F distance (when it's in the sequence)
compdist <- list()
for (i in names(fdist)) {
  # f <- fdist[[i]][,grep("E14504F", colnames(fdist[[i]]))]
  # r <- rdist[[i]][,grep("E14504F", colnames(rdist[[i]]))]
  f <- fdist[[i]][,1]
  r <- rdist[[i]][,1]
  f <- data.frame(taxon = rownames(fdist[[i]]), forward = f)
  r <- data.frame(taxon = rownames(rdist[[i]]), revcomp = r)
  compdist[[i]] <- full_join(f, r, by = "taxon")
}

# Plot the results for each alignment
for (i in names(compdist)) {
  # Melt each element of the list
  toplot <- compdist[[i]] %>%
    gather(key = "direction", value = "distance" , -taxon)
  # Plot
  ggplot(toplot, aes(x = taxon, y = distance)) +
    geom_point(aes(color = direction)) +
    labs(x = "Taxon", y = "Distance",
         title = paste("Forward vs Rev Comp alignments to E14504F: ", i, 
                       sep = "")) +
    theme(axis.text.x = element_blank())
    ggsave(paste("indiv_alignments/directionComp_", i, ".pdf", sep = ""), 
           height = 4, width = 8)
}
```

Of the regions checked, only TEF1 contained sequences that had reverse complements. I'll make a tree of just LSU and RPB2. Given the variability observed in the RPB2 region, I expect this will sufficiently resolve the tree.

Update: indeed the resolution was good, but the closest match was to Sclerotium hydrophilium, which does not contain RPB2 and in many cases does not have LSU. All of the Sclerotium in our table contain ITS, though, so I'll make sure to include that region.

```{r subset to regions of interest}
# Input regions of interest
roi <- c("LSU", "RPB2", "ITS")
roi <- names(alns) %in% roi

# subset alignment
alns <- alns[roi]
```

```{r remove gaps}
# Remove gaps present in 30% of the columns
rmgaps <- lapply(alns, function(x) del.colgapsonly(x, threshold = 0.3))
names(rmgaps) <- names(alns)
# Write alignments to files for visualization in Jalview
for (i in names(rmgaps)) {
  ape::write.dna(rmgaps[[i]], format = "fasta", file = paste("indiv_alignments/combined_", i, ".afa", sep = ""))
}
```

```{r visualize alignments, message=FALSE}
lapply(rmgaps, function(x) image.DNAbin(x, show.labels = FALSE))
```

```{r substitution models}
# ALERT! This section takes about 10 min to run
# Get alignment into class pml or phyDat as list inphydat
inphydat <- list()
for (i in names(rmgaps)) {
  temp <- tempfile()
  ape::write.dna(rmgaps[[i]], format = "fasta", file = temp)
  inphydat[[i]] <- phangorn::read.phyDat(temp, format = "fasta")
}

# Check all substitution models for each region
tests <- lapply(inphydat, function(x) phangorn::modelTest(x, model = "all"))

# Retrieve the best models
bestmodel <- sapply(tests, function(x) x$Model[which.max(x$logLik)])
names(bestmodel) <- names(rmgaps)
```

```{r individual locus trees}
# Convert to phyDat format
tophydat <- lapply(rmgaps, phyDat)

# Create a distance matrix
onetree <- lapply(tophydat, dist.ml)

# Create neighbor-joining tree
ot <- lapply(onetree, NJ)

# Maximum Likelihood on the NJ tree with the alignment
ot.ml <- list()
for (i in names(ot)) {
    ot.ml[[i]] <- pml(ot[[i]], tophydat[[i]])
}

# Apply GTR+G model
optim.ml <- lapply(ot.ml, function(x) optim.pml(x, model = "GTR", 
                                                rearrangement = "stochastic", 
                                                optGamma = TRUE))

# Bootstrap replicates
bs <- lapply(optim.ml, function(x) bootstrap.pml(x, bs = 100, optNni = TRUE, 
                                                 multicore = TRUE, 
                                                 control = pml.control(trace = 0)))
# Make into tree plotting file
trees <- list()
for (i in names(ot)) {
  trees[[i]] <- plotBS(optim.ml[[i]]$tree, bs[[i]], p = 20, type = "p")
}

# If root taxa are present, reroot, else make an unrooted tree
# There's some funky xlimit spacing here - I couldn't figure out a way to make 
# it scale with the branch lengths or something in the tree object
PlotSingleLocusTree <- function(tree, i) {
  p <- ggtree(tree) +
    geom_tiplab(size = 3) +
    geom_text2(aes(subset = !isTip, label = label), 
               size = 2, hjust = 1.1, vjust = -0.5) +
    labs(title = paste("E14505F single locus tree: ", i, sep = ""))
    if (sum(tree$edge.length) > 1) {
      p <- p + xlim_tree(1.1)
    } else {
      p <- p + xlim_tree(0.3)
    }
  p <- p + ggsave(paste("singleLocusTree_", i, ".pdf", sep = ""),
                  height = 9, width = 6.5) 
}

for (i in names(trees)) {
  if ("Tulasnella pruinosa AFTOL610" %in% trees[[i]]$tip.label) {
    tree <- root(trees[[i]], outgroup = "Tulasnella pruinosa AFTOL610", 
                 resolve.root = FALSE, edgelabel = TRUE)
    PlotSingleLocusTree(tree, i)
  } else {
    tree <- unroot(trees[[i]])
    PlotSingleLocusTree(tree, i)
  }
}

# save(trees, file = "single-locus-trees.RData")
```

```{r interleaving}
# Get all names in all alignments
intnames <- unique(unlist(sapply(rmgaps, rownames)))

# Add a sequence of N's for taxa without that locus
allns <- list()
for (i in names(rmgaps)) {
  # Grab all names NOT in the alignment
  tmp_names <- intnames[!(intnames %in% rownames(rmgaps[[i]]))]
  # Create a list of N's for each name equal to alignment length
  tmp_seqs <-  vector(mode = "list", length = length(tmp_names))
  for (j in 1:length(tmp_seqs)) {
    tmp_seqs[[j]] <- rep("n", ncol(rmgaps[[i]]))
  }
  # Convert to DNAbin
  tmp_seqs <- ape::as.DNAbin(tmp_seqs)
  # Convert to muscle alignment matrix
  tmp_seqs <- ape::muscle(tmp_seqs)
  rownames(tmp_seqs) <- tmp_names
  # Bind to N sequences to that alignment
  allns[[i]] <- rbind(rmgaps[[i]], tmp_seqs)
}

# Bind all of the alignments into one
interleaved <- do.call("cbind", allns)
```

```{r combined tree}
# Adjust names
rownames(interleaved) <- make.names(rownames(interleaved))

# Create a new file for RAxML (find the largest existing file and add one)
ls <- list.files(pattern = ".*all_\\d+")
maxiter <- max(as.numeric(gsub(".*all_(\\d+)", "\\1", ls)))

# Call RAxML (takes ~20 min)
system({
  paste("raxmlHPC-AVX -s all.phy -m GTRGAMMA -n all_", maxiter+1, " -f a -# 100 -x 12345 -p 12345", sep = "")
})
system({
  paste("raxmlHPC-AVX -m GTRGAMMA -J MRE -z RAxML_bootstrap.all_", maxiter+1, " -n tre", maxiter+1, sep = "")
})
```

```{r final tree}
t <- read.tree(paste("RAxML_bipartitions.all_", maxiter+1, sep = ""))

int <- root(t, outgroup = "Tulasnella.pruinosa.AFTOL610", 
     resolve.root = FALSE, edgelabel = TRUE)

cleanTipLabels <- function(tiplabels) {
  x <- gsub("([a-z])\\.([a-z|A-Z])", "\\1 \\2", tiplabels)
  y <- gsub("..", ". ", x, fixed = TRUE)
  return(y)
}

int$tip.label <- cleanTipLabels(int$tip.label)

p <- ggtree(int) +
      geom_tiplab(size=2) +
      geom_text2(aes(subset = !isTip, label = label), 
             size = 2, hjust = 1.1, vjust = -0.5) +
      xlim_tree(1.1)

# Create an annotation data frame
annot <- data.frame(label = int$tip.label,
           group = "1", stringsAsFactors = FALSE)
annot[grep("E145", annot$label),] <- c("E14504F", "2")
annot$group <- as.factor(annot$group)

# Add tip label annotation
p %<+% annot + 
  geom_tiplab(aes(color = group), size = 2) +
  scale_color_manual(values = c("black", "red")) +
  geom_hilight(76, fill = "steelblue", alpha = 0.1, extend = 0.28) +
  geom_hilight(95, fill = "steelblue", alpha = 0.2, extend = 0.34) +
  geom_cladelabel(node = 76, label = "Ceratobasidium",  
                  offset = 0.28, fontsize = 4) +
  geom_cladelabel(node = 95, label = "Ceratorhiza hydrophila", 
                  offset = 0.34, fontsize = 4) +
  ggsave("all_interleaved.pdf", height = 4.75, width = 7.5)

```

# References
