---
output: 
  pdf_document:
    latex_engine: xelatex
    citation_package: natbib
    keep_tex: true
    fig_caption: true
    # latex_engine: pdflatex
    # template: svm-latex-ms.tex
title: "Taxonomic assignment of endophytic isolate E14504F"
author: Dan Spakowicz
# - name: Nneoma Adaku
#   affiliation: 
# - name: Daniel J Spakowicz
#   affiliation: Program in Computational Biology and Bioinformatics, Yale University, New Haven, CT
# - name: Scott Strobel
#   affiliation: Department of Molecular Biophysics and Biochemistry; Yale University, New Haven, CT
# - name: Faye Rogers
#   affiliation: 
# abstract: ""
# keywords: "asthma, microbiome, RNAseq, metatranscriptomics"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
spacing: double
bibliography: Nneoma.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE)

# Load the required packages
list.of.packages <- c("seqinr", "googlesheets", "tidyverse", "httpuv", "rentrez",
                      "ape", "treebase", "ips", "ggtree")
new.packages <- list.of.packages[!(list.of.packages %in% 
                                     installed.packages()[,"Package"])]

# Install missing packages
if (length(new.packages)) install.packages(new.packages, 
                                           repos = "http://cran.rstudio.com/")
# Load required packages
lapply(list.of.packages, require, character.only = TRUE)

# source("https://bioconductor.org/biocLite.R")
# biocLite("muscle")
# library(muscle)
# 
source("https://bioconductor.org/biocLite.R")
biocLite("ggtree")

# Set seed
set.seed(12345)
```

# Introduction

This is the script used to build a tree for Nneoma's fungus and assign its taxonomy.

I started by rolling through the databases from the recent review http://jcm.asm.org/content/55/4/1011.full to check if any would be useful for this project.

* BOLD systems http://v4.boldsystems.org/index.php/IDS_OpenIdEngine only have ITS identification.
* Looks like this is a good place for morphological features https://aftol.umn.edu/ and can even make a nexus file to include in the tree -- but aftol has been lost? (goes to godaddy website...)
* BROAD doesn't have an identification portal
* EZBioCloud doesn't have a fungal id section
* FungiDB is just genomics
* UNITE is just ITS
* IndexFungorum doesn't have an id search (but could be useful for morphology)
* CBS can be searched directly for LSU and there are lots of good hits. However, I'd rather find a paper that has gone through the effort of identifying isolates with multiple loci
* SILVA has an LSU search https://www.arb-silva.de/ 
** Identity: 43.61, LCA tax SILVA: None
** SSU Iden: 99.37, LCA tax. SILVA: None
* RDP http://rdp.cme.msu.edu/classifier/ 
  * E14504F-LSU     Root(100%) Fungi(100%) Basidiomycota(100%) Agaricomycetes(100%) Cantharellales(100%) Ceratobasidiaceae(100%) Thanatephorus(100%)

The RDP result is strong, with 100% confidence in the genus Thanatephorus. The CBS searches also found organisms of either Thanatephorus (telomorph) or Rhizoctonia (anamorph). This will very likely be the genus to which E14504F belongs. In addition, Nneoma and I found a few papers that deal with isolates of Rhizoctonia/Thanatephorus:

* [@gonzalez_ribosomal_2001] has a bunch of Thanatephorus isolates with genbank accession numbers for ITS and 28S, but nothing outside the genus (which is necessary to demonstrate the circumscription in this case).
* [@tupac_otero_diversity_2002] just have ITS and have several genera that were isolated from orchids. It's more orchid-centric than fungus-centric.
* [@lopez-chavez_proteomic_2016] defines a Thanatephorus isolate using ITS alone. The tree shows weak node support separating Thanatephorus from Ceratobasidium, but clearly their isolate is closest to a Thana. 
* [@gonzalez_phylogenetic_2016] does a really nice job of creating a multi-locus tree. This should be the model going forward. 

# Methods

I converted the table of genbank accession numbers from [@gonzalez_phylogenetic_2016] to a google spreadsheet. 

Here are the files that Nneoma created using the Staden package (pregap & gap). As soon as these have genbank accession numbers I'll add them to the table so that they can be pulled with the other sequences from the table and remove this code block and the merge code block.

```{r read fastas as a list of strings}
files <- list.files(path = "~/Dropbox/Rainforest project/E14504F sequences/", 
                    pattern = "*.fasta", full.names = TRUE)
ofastas <- list()
for (f in files) {
  tmp <- readLines(f)
  tmp[2:length(tmp)] <- toupper(tmp[2:length(tmp)])
  ofastas[[f]] <- tmp
}

# Adjust names of list
CleanPaths <- function(names){
  tmp <- gsub(".*E14504F? (.*)\\.fasta", "\\1", names)
  tmp <- gsub(" partial| FULL| ver2", "\\1", tmp)
  return(tmp)
}
names <- CleanPaths(names(ofastas))
names(ofastas) <- names

# Cat ITS1 and ITS2 sep by 100 N's and remove ITS2
its <- grep("ITS", names(ofastas))
ofastas[[its[1]]] <- c(ofastas[[its[1]]], paste(rep("N", 100), collapse = ""), 
                       ofastas[[its[2]]][2])
ofastas <- ofastas[-its[2]]
names(ofastas)[grep("ITS", names(ofastas))] <- "ITS"
```

```{r pull sequences from google sheets as a list of strings}
# Load table into dataframe
sheet <- gs_title("E14504F")
x <- gs_read(sheet)

# Convert hyphen-only columns to NA
x <- data.frame(apply(x, 2, function(x) gsub("^-$", NA, x)), as.is = TRUE)

# Take in a character vector of genbank accession numbers and return a fasta 
# file in ape format
RetrieveSequencesAsString <- function(charvec){
  try({
    string <- entrez_fetch(db = "nucleotide", id = charvec, rettype = "fasta")
    string <- unlist(strsplit(string, split = "\n"))
    return(string)
  }, silent = TRUE)
}

# Retrieve all sequence into a list
loci <- colnames(x[,5:11])
gfastas <- list()
for (i in loci) {
  gfastas[[i]] <- RetrieveSequencesAsString(x[,grep(i, colnames(x))])
}
# Remove those without any sequences
gfastas <- gfastas[-(which(sapply(gfastas, class) == "try-error"))]
```

```{r merge e145 fastas with genbank fastas}
bfastas <- list()
for (n in names(gfastas)){
  if (n %in% names(ofastas)) {
    bfastas[[n]] <- c(gfastas[[n]], ofastas[[which(names(ofastas) %in% n)]])
  } else {
    bfastas[[n]] <- gfastas[[n]]
  }
}
```


```{r alignment}
# Read in each string as a fasta file
fastas <- list()
for (i in 1:length(bfastas)) {
  temp <- tempfile()
  write(bfastas[[i]], temp)
  fastas[[i]] <- read.FASTA(file = temp)
}

# Align sequences
alns <- lapply(fastas, ape::muscle)

# Remove gaps present in 30% of the columns
rmgaps <- lapply(alns, function(x) del.colgapsonly(x, threshold = 0.3))
names(rmgaps) <- names(bfastas)
```

```{r visualize alignments, message=FALSE}
lapply(rmgaps, function(x) image.DNAbin(x, show.labels = FALSE))
```

I can't figure out how to add titles in `image.DNAbin()`, but the order is "ITS"  "LSU"  "RPB2" "TEF1" "ATP6". The ITS looks to have a few orgs that might be revcomps, maybe three. I'm not sure how best to identify those and replace them in this format. It looks like the N length is short by ~ 20nt.

```{r reverse complement, eval=FALSE}

```

```{r substitution models, eval=FALSE}

```

```{r individual locus trees, eval=FALSE}

```

```{r interleaving, eval=FALSE}

```

```{r combined tree, eval=FALSE}

```

## Alternative methods

* retreive Treebase file and add E14504F sequences to those alignments
** This is becoming increasingly attractive given the seemingly large fraction of reverse complements observed. Particularly now that the ITS1 and ITS2 sequences are available, all sections except atp6 would be usable. It's worth taking the time to explore how to add one more sequence onto an existing alignment in R.

```{r treebase, eval=FALSE}
# Retrieve Gonzalez study
# Retd from https://treebase.org/treebase-web/search/study/matrices.html?id=15006 on 12 May 2017
nexlines <- readLines("1_1489744745_5LOCI_Cantharelloid_Ceratobasidiaceae.nexorg")

ConvertNexusToAln <- function(nexlines){
  start <- grep("MATRIX", nexlines)+1
  end <- min(grep("END;", nexlines[start:length(nexlines)], fixed = TRUE))
  aln <- nexlines[start : ((start + end)-3)]
  aln <- gsub("^\t", ">", aln)
  aln <- unlist(strsplit(aln, split = " {2,}"))
  aln <- gsub(" ", "_", aln)
  aln <- gsub("'", "", aln)
  return(aln)
}
aln <- ConvertNexusToAln(nexlines)

# Write alignment to file
writeLines(aln, "gonzalez_phylogenetic_2016.afa")

# Concatenate E14504F fastas for alignment
# order is partition Names = 4 : ITSLSU , RPB2 , EF1a , ATP6;
SingleFasta <- function(ofasta){
  e145 <- c(ofastas[["ITS"]], ofastas[["LSU"]], 
            ofastas[["RPB2"]], ofastas[["TEF1"]])
  e145 <- e145[-grep(">", e145)]
  e145 <- c(">E14504F", e145)
  return(e145)
}
e145 <- SingleFasta(ofasta)
writeLines(e145, "e14504f_gonzalez_loci.fa")
```

```{r align with muscle}
# System call for muscle (arg isn't available in the R version yet)
system({
  "muscle -profile -in1 gonzalez_phylogenetic_2016.afa -in2 e14504f_gonzalez_loci.fa -out combined.afa"
})

# This step adds ~1k characters onto the alignment... not sure what it's 
# doing and the documentation is terrible
# system({
#   "muscle -in combined.afa -out refined.afa -refine"
# })

# Convert to nexus
newaln <- read.fasta("combined.afa")
write.nexus.data(newaln, "combined.nex", interleaved = FALSE)

# Append a mr bayes block at the end of the nexus file
x <- read.nex("combined.nex")
mrbayes(x, "combined_bayes.nex")

# Write phylip file for raxml tree
write.phy(x, "combined_raxml.phy")
# Read in phy file to file "y" in sequence 4
y <- readLines("combined_raxml.phy")
y <- gsub("y", "n", y)
# Re-write phy file
writeLines(y, "combined_raxml.phy")
```

```{r RAxML, eval=FALSE}
# Call RAxML (takes ~20 min)
system({
  "raxmlHPC-AVX -s combined_raxml.phy -m GTRGAMMA -n comb_1 -p 12345 -x 12345 -# 100"
})
system({
  "raxmlHPC-AVX -m GTRGAMMA -J MRE -z RAxML_bootstrap.comb_1 -n tre"
})

```

```{r bayes block}
# Write Mr Bayes submission file to move to Grace and run there.
fileConn <- file("mrbayes_e14504f_gonzalez.sh")
writeLines(c("#!/bin/bash", 
             "#SBATCH --ntasks=10 --nodes=1",
             "#SBATCH --time=12:00:00",
             "#SBATCH --job-name=mrbayes_e14504f",
             "#SBATCH --mail-user=daniel.spakowicz@yale.edu",
             "#SBATCH --mail-type=ALL",
             "",
             "# Produced by DS_tree.Rmd",
             "# Dan Spakowicz",
             "# 14 May 2017",
             "# Submission file for the MrBayes analysis of the E14504F tree, using [@gonzalez_phylogenetic_2016] alignment as a reference",
             "",
             "cd /project/fas/gerstein/djs88/e14504f",
             "module load Apps/MrBayes/3.2.2",
             "mb combined_bayes.nex"), fileConn)
close(fileConn)

# Causes a public key error -- have to copy into terminal
# system({
#   "scp /Users/danielspakowicz/Dropbox/Rainforest\ project/tree/combined_bayes.nex djs88@grace-next.hpc.yale.edu:/project/fas/gerstein/djs88/e14504f/"
# })
# system({
#   "scp /Users/danielspakowicz/Dropbox/Rainforest\ project/tree/mrbayes_e14504f_gonzalez.sh djs88@grace-next.hpc.yale.edu:/gpfs/home/fas/gerstein/djs88/myscripts/projects/e14504f/"
# })

```

```{r, eval=FALSE}
# WARNING : This step takes ~5 hours. Set to eval=FALSE by default
# system({
#   "mb combined_bayes.nex"
# })
```

```{r tree visualization}
z <- read.tree("RAxML_MajorityRuleExtendedConsensusTree.tre")
ggplot(z) +
  ggtree::geom_tree() +
  ggtree::geom_tiplab(size = 2)
```


After a few segmentation faults I decided to push this to grace to run there, and just have it write the slurm submission and scp the files from this script. 

## Morphological analysis
https://aftol.umn.edu/species/Thanatephorus_cucumeris 

Look for uniperforate septa with swollen margin 

Have to distinguish it from other members of Ceratobasidaceae, namely https://aftol.umn.edu/species/Ceratobasidium_cornigerum

Oh damn they're very similar. Looks like the difference is determined by electron microscopy, in the substructure of the pore cap. Namely, Ceratobasidium has a cap with uniform electron-dense contents and Thanatephorus has a cap with three internal layers (central electron-dense layer between two less electron dense layers). Well that's not super helpful. 

Directly copied from wikipedia
>> R. solani does not produce spores, hence is identified only from mycelial characteristics or DNA analysis. Its hyphal cells are multinucleated. It produces white to deep brown mycelium when grown on an artificial medium. The hyphae are 4–15 μm wide and tend to branch at right angles. A septum near each hyphal branch and a slight constriction at the branch are diagnostic. R. solani is subdivided into anastomosis groups based on hyphal fusion between compatible strains.[12][13] 

The teleomorph of R. solani is Thanatephorus cucumeris. It forms club-shaped basidia with four apical sterigmata on which oval, hyaline basidiospores are borne.

Nice pictures here http://www.apsnet.org/edcenter/intropp/lessons/fungi/Basidiomycetes/Pages/Rhizoctonia.aspx
"Rhizoctonia species can be divided into two groups based on the number of nuclei present in cells of the hyphae. This characteristic can be assessed by staining hyphae with a fluorescent dye that binds specifically to the DNA in the nucleus. Binucleate Rhizoctonia species have two nuclei in each cell (Figure 23). Rhizoctonia cerealis is the only binucleate species that causes turfgrass disease. Multinucleate species have three or more nuclei in each cell (Figure 24). Rhizoctonia solani, R. zeae, and R. oryzae are multinucleate species."

"Rhizoctonia solani forms colonies on potato dextrose agar (PDA) that range in color from buff to black (Figure 25). Sclerotia are produced on the surface of cultures after 4 to 6 weeks, and are irregularly shaped, light tan to black, and usually >1 mm (0.04 in.) in width (Figure 26). The optimum temperature for growth of R. solani in pure culture ranges from 18ºC (64.4ºF) to 28ºC (82.4ºF). This species is divided into 11 AGs. Isolates in AGs 1 and 2 are commonly associated with turfgrass disease."



# Results and Discussion

# References