\documentclass[11pt,]{article}
\usepackage[]{mathpazo}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={Taxonomic assignment of endophytic isolate E14504F},
            pdfauthor={Dan Spakowicz},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{natbib}
\bibliographystyle{plainnat}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{{#1}}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{{#1}}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{{#1}}}
\newcommand{\ImportTok}[1]{{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{{#1}}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{{#1}}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{{#1}}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{{#1}}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{{#1}}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{{#1}}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{{#1}}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{{#1}}}}
\newcommand{\BuiltInTok}[1]{{#1}}
\newcommand{\ExtensionTok}[1]{{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{{#1}}}
\newcommand{\RegionMarkerTok}[1]{{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{{#1}}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{{#1}}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{{#1}}}}
\newcommand{\NormalTok}[1]{{#1}}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}
  \title{Taxonomic assignment of endophytic isolate E14504F}
  \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
  \author{Dan Spakowicz}
  \preauthor{\centering\large\emph}
  \postauthor{\par}
  \predate{\centering\large\emph}
  \postdate{\par}
  \date{May 10, 2017}


\begin{document}
\maketitle

\section{Introduction}\label{introduction}

This is the script used to build a tree for Nneoma's fungus and assign
its taxonomy.

I started by rolling through the databases from the recent review
\url{http://jcm.asm.org/content/55/4/1011.full} to check if any would be
useful for this project.

\begin{itemize}
\tightlist
\item
  BOLD systems
  \url{http://v4.boldsystems.org/index.php/IDS_OpenIdEngine} only have
  ITS identification.
\item
  Looks like this is a good place for morphological features
  \url{https://aftol.umn.edu/} and can even make a nexus file to include
  in the tree -- but aftol has been lost? (goes to godaddy
  website\ldots{})
\item
  BROAD doesn't have an identification portal
\item
  EZBioCloud doesn't have a fungal id section
\item
  FungiDB is just genomics
\item
  UNITE is just ITS
\item
  IndexFungorum doesn't have an id search (but could be useful for
  morphology)
\item
  CBS can be searched directly for LSU and there are lots of good hits.
  However, I'd rather find a paper that has gone through the effort of
  identifying isolates with multiple loci
\item
  SILVA has an LSU search \url{https://www.arb-silva.de/}
\item
  Identity: 43.61, LCA tax SILVA: None
\item
  SSU Iden: 99.37, LCA tax. SILVA: None
\item
  RDP \url{http://rdp.cme.msu.edu/classifier/}
\item
  E14504F-LSU Root(100\%) Fungi(100\%) Basidiomycota(100\%)
  Agaricomycetes(100\%) Cantharellales(100\%) Ceratobasidiaceae(100\%)
  Thanatephorus(100\%)
\end{itemize}

The RDP result is strong, with 100\% confidence in the genus
Thanatephorus. The CBS searches also found organisms of either
Thanatephorus (telomorph) or Rhizoctonia (anamorph). This will very
likely be the genus to which E14504F belongs. In addition, Nneoma and I
found a few papers that deal with isolates of Rhizoctonia/Thanatephorus:

\begin{itemize}
\tightlist
\item
  \citep{gonzalez_ribosomal_2001} has a bunch of Thanatephorus isolates
  with genbank accession numbers for ITS and 28S, but nothing outside
  the genus (which is necessary to demonstrate the circumscription in
  this case).
\item
  \citep{tupac_otero_diversity_2002} just have ITS and have several
  genera that were isolated from orchids. It's more orchid-centric than
  fungus-centric.
\item
  \citep{lopez-chavez_proteomic_2016} defines a Thanatephorus isolate
  using ITS alone. The tree shows weak node support separating
  Thanatephorus from Ceratobasidium, but clearly their isolate is
  closest to a Thana.
\item
  \citep{gonzalez_phylogenetic_2016} does a really nice job of creating
  a multi-locus tree. This should be the model going forward.
\end{itemize}

\section{Methods}\label{methods}

I converted the table of genbank accession numbers from
\citep{gonzalez_phylogenetic_2016} to a google spreadsheet.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Load table into dataframe}
\NormalTok{sheet <-}\StringTok{ }\KeywordTok{gs_title}\NormalTok{(}\StringTok{"E14504F"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Sheet successfully identified: "E14504F"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{x <-}\StringTok{ }\KeywordTok{gs_read}\NormalTok{(sheet)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Accessing worksheet titled 'Sheet1'.
\end{verbatim}

\begin{verbatim}
## Parsed with column specification:
## cols(
##   Name = col_character(),
##   `Voucher ID` = col_character(),
##   Host = col_character(),
##   `Geographic Origin` = col_character(),
##   ITS = col_character(),
##   SSU = col_character(),
##   LSU = col_character(),
##   RPB1 = col_character(),
##   RPB2 = col_character(),
##   TEF1 = col_character(),
##   ATP6 = col_character(),
##   Reference = col_character(),
##   Notes = col_character()
## )
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Convert hyphen-only columns to NA}
\NormalTok{x <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{apply}\NormalTok{(x, }\DecValTok{2}\NormalTok{, function(x) }\KeywordTok{gsub}\NormalTok{(}\StringTok{"^-$"}\NormalTok{, }\OtherTok{NA}\NormalTok{, x)), }\DataTypeTok{as.is =} \OtherTok{TRUE}\NormalTok{)}

\CommentTok{# Take in a character vector of genbank accession numbers and return a fasta file in ape format}
\NormalTok{RetrieveSequences <-}\StringTok{ }\NormalTok{function(charvec)\{}
  \KeywordTok{try}\NormalTok{(\{}
    \NormalTok{string <-}\StringTok{ }\KeywordTok{entrez_fetch}\NormalTok{(}\DataTypeTok{db =} \StringTok{"nucleotide"}\NormalTok{, }\DataTypeTok{id =} \NormalTok{charvec, }\DataTypeTok{rettype =} \StringTok{"fasta"}\NormalTok{)}
    \NormalTok{lsu <-}\StringTok{ }\KeywordTok{unlist}\NormalTok{(}\KeywordTok{strsplit}\NormalTok{(string, }\DataTypeTok{split =} \StringTok{"}\CharTok{\textbackslash{}n}\StringTok{"}\NormalTok{))}
    \NormalTok{temp <-}\StringTok{ }\KeywordTok{tempfile}\NormalTok{()}
    \KeywordTok{write}\NormalTok{(lsu, temp)}
    \NormalTok{lsu <-}\StringTok{ }\NormalTok{ape::}\KeywordTok{read.dna}\NormalTok{(temp, }\DataTypeTok{format =} \StringTok{"fasta"}\NormalTok{)}
    \KeywordTok{return}\NormalTok{(lsu)}
  \NormalTok{\}, }\DataTypeTok{silent =} \OtherTok{TRUE}\NormalTok{)}
\NormalTok{\}}

\CommentTok{# Retrieve all sequence into a list}
\NormalTok{loci <-}\StringTok{ }\KeywordTok{colnames}\NormalTok{(x[,}\DecValTok{5}\NormalTok{:}\DecValTok{11}\NormalTok{])}
\NormalTok{fastas <-}\StringTok{ }\KeywordTok{list}\NormalTok{()}
\NormalTok{for (i in loci) \{}
  \NormalTok{fastas[[i]] <-}\StringTok{ }\KeywordTok{RetrieveSequences}\NormalTok{(x[,}\KeywordTok{grep}\NormalTok{(i, }\KeywordTok{colnames}\NormalTok{(x))])}
\NormalTok{\}}
\CommentTok{# Remove those without any sequences}
\NormalTok{fastas <-}\StringTok{ }\NormalTok{fastas[-(}\KeywordTok{which}\NormalTok{(}\KeywordTok{sapply}\NormalTok{(fastas, class) ==}\StringTok{ "try-error"}\NormalTok{))]}
\end{Highlighting}
\end{Shaded}

Here are the files that Nneoma created using the Staden package (pregap
\& gap). As soon as these have genbank accession numbers I'll add them
to the table so that they can be pulled with the other sequences from
the table.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Read in Nneoma's files}
\NormalTok{lsu <-}\StringTok{ }\KeywordTok{read.dna}\NormalTok{(}\DataTypeTok{file =} \StringTok{"~/Dropbox/Rainforest project/E14504F sequences/E14504F LSU FULL.fasta"}\NormalTok{, }\DataTypeTok{format =} \StringTok{"fasta"}\NormalTok{)}
\CommentTok{# rpb1 <- read.dna(file = "~/Dropbox/Rainforest project/E14504F sequences/E14504 RPB1.fasta")}
\NormalTok{rpb2 <-}\StringTok{ }\KeywordTok{read.dna}\NormalTok{(}\DataTypeTok{file =} \StringTok{"~/Dropbox/Rainforest project/E14504F sequences/E14504F RPB2.fasta"}\NormalTok{, }\DataTypeTok{format =} \StringTok{"fasta"}\NormalTok{)}
\CommentTok{# ssu <- read.dna(file = "~/Dropbox/Rainforest project/E14504F sequences/E14504F SSU ver2.fasta")}
\NormalTok{tef1 <-}\StringTok{ }\KeywordTok{read.dna}\NormalTok{(}\DataTypeTok{file =} \StringTok{"~/Dropbox/Rainforest project/E14504F sequences/E14504F TEF1.fasta"}\NormalTok{, }\DataTypeTok{format =} \StringTok{"fasta"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{# Align sequences}
\NormalTok{alns <-}\StringTok{ }\KeywordTok{lapply}\NormalTok{(fastas, ape::muscle)}
\end{Highlighting}
\end{Shaded}

\section{Results and Discussion}\label{results-and-discussion}

\renewcommand\refname{References}
\bibliography{Nneoma.bib}


\end{document}
